pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1050,15 +1043,15 @@ void messageReceived(MessageIdData messageId, int redeliveryCount, List<Long> ac
 
             // Enqueue the message so that it can be retrieved when application calls receive()
             // if the conf.getReceiverQueueSize() is 0 then discard message if no one is waiting for it.
-            // if asyncReceive is waiting then notify callback without adding to incomingMessages queue
             if (deadLetterPolicy != null && possibleSendToDeadLetterTopicMessages != null &&
                     redeliveryCount >= deadLetterPolicy.getMaxRedeliverCount()) {
                 possibleSendToDeadLetterTopicMessages.put((MessageIdImpl)message.getMessageId(),
                         Collections.singletonList(message));
             }
+            enqueueMessage(message);
             if (peekPendingReceive() != null) {
-                notifyPendingReceivedCallback(message, null);
-            } else if (enqueueMessageAndCheckBatchReceive(message) && hasPendingBatchReceive()) {
+                notifyPendingReceivedCallback(incomingMessages.poll(), null);
```

Comment:
If it is to be lock-free, I suggest to use executorProvider, the logic of the same consumerId will be processed in the same thread. Then compare the impact of performance.

If you want to lock, we only need to ensure that the thread safety of these two scenarios can reduce the scope of the lock.
1) When the poll comes out and the message is empty, put it into the pendingQueue
2) Judge the pendingQueue is empty, enter enqueueMessageAndCheckBatchReceive

Your current logic still has thread safety issues. If only one thread is consuming, it may cause a message in the queue to never be consumed, and there is also a pending future in the pendingQueue. This situation will occur in the scene of the last message.

Reply:
> Your current logic still has thread safety issues. If only one thread is consuming, it may cause a message in the queue to never be consumed, and there is also a pending future in the pendingQueue. This situation will occur in the scene of the last message.

Thanks for your review, could you show this case as time line?

Reply:
notify thread is thread-1
Consumer Receive thread is thread-2

1 thread-2 poll out of the message is empty
2 thread-1 The message enters the queue and judges whether the pending is empty, and it is found to be empty
3 thread-2 is put into pending future (not batch Receive)

Reply:
Thanks, I see.
What about adding heartbeat message to avoid thread being hang up? 
I attend to lock-free solution.

Reply:
In my opinion, it is not a good way. It changes the current behavior and makes the receiving of messages less timely.